library/waku_thread/inter_thread_communication/node_lifecycle_request.nim

```diff
@@ -6,32 +6,118 @@ import
+  ../../../waku/node/config,
+  ../../../waku/waku_relay/protocol,
+  ../../events/[json_error_event,json_message_event,json_base_event],
+  ../config,
   ./request
 
 type
   NodeLifecycleMsgType* = enum
+    CREATE_NODE
     START_NODE
     STOP_NODE
 
 type
   NodeLifecycleRequest* = ref object of InterThreadRequest
     operation: NodeLifecycleMsgType
+    configJson: cstring ## Only used in 'CREATE_NODE' operation
 
 proc new*(T: type NodeLifecycleRequest,
-          op: NodeLifecycleMsgType): T =
+          op: NodeLifecycleMsgType,
+          configJson: cstring = ""): T =
 
-  return NodeLifecycleRequest(operation: op)
+  return NodeLifecycleRequest(operation: op, configJson: configJson)
+
+proc createNode(configJson: cstring): Result[WakuNode, string] =
+  var privateKey: PrivateKey
+  var netConfig = NetConfig.init(ValidIpAddress.init("127.0.0.1"),
+                                 Port(60000'u16)).value
+  var relay: bool
+  var topics = @[""]
+  var jsonResp: JsonEvent
+
+  if not parseConfig($configJson,
+                     privateKey,
+                     netConfig,
+                     relay,
+                     topics,
+                     jsonResp):
+    return err($jsonResp)
+
+  var enrBuilder = EnrBuilder.init(privateKey)
+
+  enrBuilder.withIpAddressAndPorts(
+    netConfig.enrIp,
+    netConfig.enrPort,
+    netConfig.discv5UdpPort
+  )
+
+  if netConfig.wakuFlags.isSome():
+    enrBuilder.withWakuCapabilities(netConfig.wakuFlags.get())
+
+  enrBuilder.withMultiaddrs(netConfig.enrMultiaddrs)
+
+  let addShardedTopics = enrBuilder.withShardedTopics(topics)
+  if addShardedTopics.isErr():
+    let msg = "Error setting shared topics: " & $addShardedTopics.error
+    return err($JsonErrorEvent.new(msg))
+
+  let recordRes = enrBuilder.build()
+  let record =
+    if recordRes.isErr():
+      let msg = "Error building enr record: " & $recordRes.error
+      return err($JsonErrorEvent.new(msg))
+
+    else: recordRes.get()
+
+  var builder = WakuNodeBuilder.init()
+  builder.withRng(crypto.newRng())
+  builder.withNodeKey(privateKey)
+  builder.withRecord(record)
+  builder.withNetworkConfiguration(netConfig)
+  builder.withSwitchConfiguration(
+    maxConnections = some(50.int)
+  )
+
+  let wakuNodeRes = builder.build()
+  if wakuNodeRes.isErr():
+    let errorMsg = "failed to create waku node instance: " & wakuNodeRes.error
+    return err($JsonErrorEvent.new(errorMsg))
+
+  var newNode = wakuNodeRes.get()
+
+  if relay:
+    waitFor newNode.mountRelay()
+    newNode.peerManager.start()
+
+  return ok(newNode)
 
 method process*(self: NodeLifecycleRequest,
-                node: WakuNode): Future[Result[string, string]] {.async.} =
+                node: ptr WakuNode): Future[Result[string, string]] {.async.} =
 
   case self.operation:
+    of CREATE_NODE:
+      let newNodeRes = createNode(self.configJson)
+      if newNodeRes.isErr():
+        return err(newNodeRes.error)
+
+      node[] = newNodeRes.get()
```

Comment:
What does this notation (the `[]`) mean/do?

Reply:
Thanks for the comments! I'll fix them asap.
This `[]` is used for `ptr` types. I understand the `[]` operator as a way to dereference the ptr, i.e. a way to access the value or object it points to.

Something interesting I've learned recently:

| ref obj A | ptr A |
|--------|--------|
| Points to data stored in the heap | Points to data stored either in the heap or stack |
| GC'ed | The developer should take care of allocating and deallocating manually |
| No thread-safe | Thread-safe |

Reply:
Thank you very much!